### DISCUSSION TITLE
Controller rendered input field that returns File objects doesn't ever dirty

### DISCUSSION THREAD

**Discussion by xchau** [react-hook-form/react-hook-form#3805] [Q&A] [answered]
I'm using Shopify Polaris's `DropZone` component, which allows images to be dropped/dragged onto it. The returned params once dropped/dragged is `File[]`, which `onChange` handles great. The issue is, once changed, `formState` does not register this controlled component as being dirty. 

 I've tried bypassing `onChange` and passing a custom handler that invokes `setValue("group.myImages", images, { shouldDirty: true })` as well, but that also does not work. However, I do notice that if I pass simple primitives like an array of strings/normal objects, it works. Am I missing a step or is this a bug?

CSB: https://codesandbox.io/s/httpsgithubcomreact-hook-formreact-hook-formdiscussions3805-60nzb

*Form:*
```
<form onSubmit={handleSubmit(onSubmit)} ... />
``` 

*Handler:*
```
const onSubmit = (data: any) => {
    console.log("the data:", data); // data is present
    console.log("form is dirty?", formState.isDirty); // false 
    console.log("form has dirty fields:", formState.dirtyFields); // {}
}
```

*DropZone:*
```
<Controller
    name="group.myImages"
    control={control}
    defaultValue={[]}
    render={({ onChange }) => {
        return (
            <DropZone
                id="my-dropzone"
                key="my-dropzone"
                accept="image/jpeg,image/jpg,image/png"
                onDrop={onChange}
                type="image"
                label="an image"
              />
          )
     }}
/>
```

**Comment by bluebill1049** [comment, marked as answer]
provide a csb.

**Reply by xchau** [reply]
Sure, here you go: https://codesandbox.io/s/httpsgithubcomreact-hook-formreact-hook-formdiscussions3805-60nzb

**Reply by xchau** [reply]
It seems the first upload doesn't trigger the "dirtying" of the field, but the second time does.

**Reply by bluebill1049** [reply]
Thanks for the CSB. i see, it's for the dirty state, it's a tricky one. 

> File typed input will need to manage at app level due to the ability to cancel file selection and FileList object.

<img width="987" alt="Screen Shot 2020-12-30 at 12 46 37 pm" src="https://user-images.githubusercontent.com/10513364/103324822-0fc7b800-4a9d-11eb-97e9-711e3039ce0e.png">

you may want to manually setValue with shouldDirty to trigger the dirty.

**Reply by xchau** [reply]
Ah, thanks. To clarify I have tried using `setValue` & setting `shouldDirty` but that also yields a mixed result:
```
<DropZone
    id="my-dropzone"
    key="my-dropzone"
    accept="image/jpeg,image/jpg,image/png"
    onDrop={(files) => {
        setValue("group.myImages", files, { shouldDirty: true });
    }}
    type="image"
    label="an image"
/>
```

So I guess what you're suggesting is to use a state variable to track dirty state? 👍

**Reply by bluebill1049** [reply]
yes that's what I meant 👍

**Comment by xchau** [comment]
Thanks for the help @bluebill1049

**Reply by bluebill1049** [reply]
np :)
